### DISCUSSION TITLE
Separating all secrets from the Docker build process

### DISCUSSION THREAD

**Discussion by jplauri** [apache/airflow#20668] [Q&A]
I'm trying to run Airflow via Docker in a corporate environment. As per the company policy, any images such as Airflow should be built outside the production environment where various security checks will be performed. Moreover, the built images will be accessible by everyone and should not contain any secrets such as Airflow backend credentials. 

My Airflow build is simple, I'm essentially using the Airflow [docker-compose.yaml](https://airflow.apache.org/docs/apache-airflow/stable/docker-compose.yaml) where I'm specifying a backend database (i.e., setting AIRFLOW__CORE__SQL_ALCHEMY_CONN) and mounting some network drives.

Just to test things locally, I commented out AIRFLOW__CORE__SQL_ALCHEMY_CONN from the environment part of the yaml file with the idea of simulating the way I must build Airflow without any secrets & hoping I could later on "at runtime" set the backend credentials. However, what happens is essentially the following errors upon executing "docker-compose up":

```
ERROR: for airflow-webserver  Container "3feb811cbcb6" exited with code 1.

ERROR: for airflow-triggerer  Container "3feb811cbcb6" exited with code 1.

ERROR: for flower  Container "3feb811cbcb6" exited with code 1.

ERROR: for airflow-worker  Container "3feb811cbcb6" exited with code 1.

ERROR: for airflow-scheduler  Container "3feb811cbcb6" exited with code 1.
ERROR: Encountered errors while bringing up the project.
```

If I change AIRFLOW__CORE__SQL_ALCHEMY_CONN to the default "postgresql+psycopg2://airflow:airflow@postgres/airflow" things work fine.

I was able to find [this answer on SO](https://stackoverflow.com/questions/68714224/airflow-exiting-after-initilalization) but I don't think I wasn't able to parse much from it. 

Really, my question is, is it possible to achieve what I want, i.e., separate all secrets from the image building process? If so, is what I'm trying to do the right way (roughly) of going about it?

**Comment by jplauri** [comment]
Oops, seems like I am unable to delete this myself. Could someone do it for me? I believe the answer is quite simple: it's possible to provide the secrets in an .env file.

**Reply by potiuk** [reply]
Actually - you can get some answers here: https://airflow.apache.org/docs/docker-stack/build.html#using-custom-installation-sources - adding env is not necessary best, but Airlfow uses multi-segmented build with built-in capabilities of some customisations specifically targetted for corporate environment and secrets as well. Also I am just about to convert the build process of the image to use BUILDKIT - PR [here](https://github.com/apache/airflow/pull/20664) which has even more capabilities https://docs.docker.com/develop/develop-images/build_enhancements/#new-docker-build-secret-information

**Reply by potiuk** [reply]
So the question was good - you should consider restoring it :).

**Reply by jplauri** [reply]
Excellent, thank you! :-) I rolled back the original question. It appears that using an .env works, but somehow I missed those parts in the docs. I will check them out. Hopefully this is useful for someone else too.
